DAOS-10625 control: Create the tool to collect the logs/config for support purpose

ravalsam opened the pull request:
1> Adding dmg support collectlog option collect the logs/configs from servers. 
2> Adding daos_server support collectlog option to get the system and server side information in case dmg Management layer is not working for some reason.
3> Adding daos_admin support collectlog option to get the client side logs and configs.
4> Added unit test for support lib
5> Added functional tests for dmg,daos_agent and daos_server support options

Signed-off-by: Samir Raval <samir.raval@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/1/execution/node/1092/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-11094/3/testReport/(root)/

tanabarr commented on src/control/cmd/dmg/support.go:
```diff
@@ -0,0 +1,182 @@
+			}
+			if len(resp.GetHostErrors()) > 0 {
+				var bld strings.Builder
+				_ = pretty.PrintResponseErrors(resp, &bld)
```
errors should always be handled

tanabarr commented on src/control/cmd/dmg/support.go:
```diff
@@ -0,0 +1,182 @@
+	}
+
+	// Rsync the logs from servers
+	hostName, _ := support.GetHostName()
```
errors should always be handled

tanabarr reviewed (changes requested):
some initial observations (most points refer to multiple occurrences in the PR):

if b == true can be shortened to if b

initialising to an empty value i.e. LogCollection["..."] = []string{""} it is rarely necessary as zero values are usually good enough un-initialised (one of the useful things of go)

If using string constants as keys, define them e.g. const CopyServerConfigCmd = "CopyServerConfig", not convinced this is the best use of the map though, you might be able to define enum for the keys

progress.Total = progress.Total + 1 can be shortened to progress.Total++

The code for the daos_server and dmg command variants should be consolidated into shared helper functions to reduce duplication.

kjacque commented on src/control/cmd/daos_agent/support.go:
```diff
@@ -0,0 +1,92 @@
+			return err
+		}
+
+		for i := 1; i < 3; i++ {
+			os.RemoveAll(cmd.TargetFolder)
+		}
```
Why would you need to do this twice to the same folder?

kjacque commented on src/control/cmd/daos_server/support.go:
```diff
@@ -0,0 +1,94 @@
+	}
+
+	// Default 4 steps of log/conf collection.
+	progress := support.ProgressBar{1, 4, 0, false}
```
The magic numbers are very unclear to me. Are these derived from the LogCollection map length? If so you can use `len(LogCollection)`

kjacque commented on src/control/cmd/daos_server/support.go:
```diff
@@ -0,0 +1,94 @@
+	CustomLogs   string `short:"c" long:"custom-logs" description:"Collect the Logs from given directory"`
+}
+
+func (cmd *collectLogCmd) Execute(_ []string) error {
```
So much of this function seems identical to the one in the agent. Could most of this be refactored into a CollectAllLogs function in the new `lib/support` package you created? Seems like it takes the params above, plus the LogCollection steps.

kjacque commented on src/control/cmd/dmg/support.go:
```diff
@@ -0,0 +1,182 @@
+		"CollectDaosServerCmd": support.DaosServerCmd,
+	}
+
+	// Default 7 set of support collection steps to show in progress bar
+	progress := support.ProgressBar{1, 7, 0, cmd.jsonOutputEnabled()}
```
Not a fan of the magic numbers again. This time I'm not sure where the 7 comes from.

kjacque commented on src/control/cmd/dmg/support.go:
```diff
@@ -0,0 +1,182 @@
+		support.PrintProgress(&progress)
+	}
+
+	// Rsync the logs from servers
```
For readability and maintainability, it would be better to break some of these subsections out into helper methods.

ravalsam replied:
I still need to work on better rsync logic so I will update helper method with final code.

kjacque commented on src/control/cmd/dmg/support_test.go:
```diff
@@ -0,0 +1,28 @@
+			"support collectlog without any args",
+			"support collectlog",
+			printRequest(t, &control.CollectLogReq{}),
+			errors.New("DAOS Management Service is down"),
```
This is a very strange expectation!

kjacque commented on src/control/cmd/dmg/support_test.go:
```diff
@@ -0,0 +1,28 @@
+
+	"github.com/pkg/errors"
+
+	// "github.com/daos-stack/daos/src/control/cmd/dmg/pretty"
+	"github.com/daos-stack/daos/src/control/lib/control"
+	// "github.com/daos-stack/daos/src/control/lib/support"
```
Don't forget to remove these commented out lines

kjacque commented on src/control/common/archive.go:
```diff
@@ -0,0 +1,61 @@
+	"os"
+	"path/filepath"
+)
+
+func FolderCompress(src string, buf io.Writer) error {
```
Please add a documentation comment for this function.

kjacque commented on src/control/common/file_utils.go:
```diff
@@ -257,3 +258,74 @@ func FindBinary(binName string) (string, error) {
 	return adjPath, nil
 }
+
+// Copy file from src to dst.
```
These comments should start with the name of the function being documented, and be in the form of a complete sentence. For example something like "CpFile copies a file from src to dst." The comments are used to autogenerate documentation.

kjacque commented on src/control/common/file_utils.go:
```diff
@@ -257,3 +258,74 @@ func FindBinary(binName string) (string, error) {
+func IsHidden(filename string) bool {
+	if filename[0:1] == "." {
+		return true
+	} else {
+		return false
+	}
```
don't need this else block since we return false below

kjacque commented on src/control/lib/support/log.go:
```diff
@@ -0,0 +1,640 @@
+
+// Folder names to copy logs and configs
+const (
+	dmgSystemLogFolder = "DmgSystemLog"     // Copy the dmg command output for DAOS system
+	daosNodeLogFolder  = "DaosNodeLog"      // Copy the dmg command output specific to the storage.
+	daosAgentCmdInfo   = "DaosAgentCmdInfo" // Copy the daos_agent command output specific to the node.
+	systemInfo         = "SysInfo"          // Copy the system related information
+	serverLogs         = "ServerLogs"       // Copy the server/control and helper logs
+	clientLogs         = "ClientLogs"       // Copy the server/control and helper logs
+	daosConfig         = "ServerConfig"     // Copy the server config
+	customLogs         = "CustomLogs"       // Copy the Custom logs
```
If these are all directory names, please be consistent about the variable naming.

kjacque commented on src/control/lib/support/log.go:
```diff
@@ -0,0 +1,640 @@
+	Start      int  // start int number
+	Total      int  // end int number
+	Steps      int  // Int number be increased per steps
+	JsonOutput bool // Json option to skip progress bar if it's enabled
```
name suggestion - "NoDisplay"

kjacque commented on src/control/lib/support/log.go:
```diff
@@ -0,0 +1,640 @@
+	JsonOutput bool // Json option to skip progress bar if it's enabled
+}
+
+type Params struct {
```
name suggestion - CollectLogsParams

Considering we may someday add other support commands.

kjacque commented on src/control/lib/support/log.go:
```diff
@@ -0,0 +1,640 @@
+	LogCmd       string
+}
+
+type copy struct {
```
copy of what? I'd suggest a different name.

kjacque commented on src/control/lib/support/log.go:
```diff
@@ -0,0 +1,640 @@
+}
+
+// Print the progress bar during log collect command
+func PrintProgress(progBar *ProgressBar) {
```
suggestion - could make these member functions of ProgressBar.A function signature like:

`func (p *ProgressBar) Display(out io.Writer)`

It would be easy enough to decide within the function whether to print progress or end status.

kjacque commented on src/control/lib/support/log.go:
```diff
@@ -0,0 +1,640 @@
+func CollectSupportLog(log logging.Logger, opts ...Params) error {
+	switch opts[0].LogFunction {
+	case "CopyServerConfig":
+		return CopyServerConfig(log, opts...)
```
Do all of these functions need to be exported? Starting with a capital letter in Go means the symbol is exported. If they are solely called from this function I'd suggest unexporting them.

kjacque reviewed (changes requested):
A lot of code! Most of my comments are Go language tips and/or naming suggestions, and can be applied in multiple places.

I haven't tried running it so can't speak to any oddities there.

ravalsam commented:
Thank you and appreciated for the review comments. I will update the PR.

daosbuild1 commented on src/tests/ftest/control/dmg_support_collectlog.py:
```diff
@@ -0,0 +1,27 @@
+        :avocado: tags=hw,medium
+        :avocado: tags=basic,control,dmg
+        :avocado: tags=test_dmg_support_collect_log
+        """        
```
(style)  trailing whitespace

daosbuild1 replied:
```suggestion
        """
```

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-11094/5/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/5/execution/node/145/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/12/execution/node/1211/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/12/execution/node/1237/log

daosbuild1 commented on src/tests/ftest/util/support_test_base.py:
```diff
@@ -0,0 +1,142 @@
+        return None
+
+    def cleanup_support_log(self, log_dir):
+        """ Test cleanup to remove the temporary directory  
```
(style)  trailing whitespace

daosbuild1 replied:
```suggestion
        """ Test cleanup to remove the temporary directory
```

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-11094/13/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/13/execution/node/145/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/21/execution/node/1015/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/21/execution/node/1194/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/22/execution/node/176/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/22/execution/node/220/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.4 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/22/execution/node/217/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/23/execution/node/513/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/26/execution/node/1014/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/26/execution/node/1193/log

ravalsam commented:
All unit tests and functional test passed in Ci.

There are Test failure not because of this PR, but an existing defect.

- control/daos_object_query.py - DAOS-13060
- control/daos_snapshot.py - DAOS-13060
- osa/online_drain.py - DAOS-13054

daltonbohning commented on src/tests/ftest/control/daos_agent_support_collect_log.py:
```diff
@@ -0,0 +1,53 @@
+        :avocado: tags=all,full_regression
+        :avocado: tags=hw,medium
+        :avocado: tags=control,basic,support,daos_server
+        :avocado: tags=test_daos_agent_support_collect_log
```
Please add the class name tag as well
```suggestion
        :avocado: tags=DaosAgentSupportCollectLogTest,test_daos_agent_support_collect_log
```

daltonbohning commented on src/tests/ftest/control/daos_server_support_collect_log.py:
```diff
@@ -0,0 +1,57 @@
+        :avocado: tags=all,daily_regression
+        :avocado: tags=hw,medium
+        :avocado: tags=control,basic,support,daos_server
+        :avocado: tags=test_daos_server_support_collect_log
```
```suggestion
        :avocado: tags=DaosSupportCollectLogTest,test_daos_server_support_collect_log
```

daltonbohning commented on src/tests/ftest/control/dmg_support_collect_log.py:
```diff
@@ -0,0 +1,55 @@
+        :avocado: tags=all,daily_regression
+        :avocado: tags=hw,medium
+        :avocado: tags=basic,control,support,dmg
+        :avocado: tags=test_dmg_support_collect_log
```
```suggestion
        :avocado: tags=DmgSupportCollectLogTest,test_dmg_support_collect_log
```

ravalsam replied:
Done

daltonbohning commented on src/tests/ftest/control/dmg_support_collect_log.py:
```diff
@@ -0,0 +1,55 @@
+                                              target_folder=self.target_folder,
+                                              archive=True)
+        status = result["status"]
+        self.assertEqual(status, 0, "bad return status")
```
Recommend more descriptive error message
```suggestion
        self.assertEqual(status, 0, "bad return status from dmg support collect-log")
```

daltonbohning commented on src/tests/ftest/util/agent_utils.py:
```diff
@@ -148,6 +178,23 @@ def dump_attachinfo(self, output="uri.txt"):
+
+        """
+        self.set_command(("support", "collect-log"), **kwargs)
+        return self._get_result()
```
Any reason to not use `self._get_json_result` here? Phil has been updating almost all functions to use json by default

daltonbohning commented on src/tests/ftest/util/agent_utils.py:
```diff
@@ -237,6 +284,30 @@ def dump_attachinfo(self):
+        Raises:
+            CommandFailure: if the daos_agent command fails.
+
+        Returns:
+            dict: the daos_agent command output converted to a python dictionary
```
This returns `run_remote(...)`
```suggestion
        Returns:
            RemoteCommandResult: a grouping of the command results from the same hosts with the same return status
```

daltonbohning commented on src/tests/ftest/util/dmg_utils.py:
```diff
@@ -512,6 +512,42 @@ def server_set_logmasks(self, masks=None, raise_exception=None):
+            dict: the dmg json command output converted to a python dictionary
+
+        """
+        # Example JSON output:
+        # {
+        #   "response": {
+        #     "host_errors": {}
+        #   },
+        #   "error": null,
+        #   "status": 0
+        # }
```
I'd recommend removing this "Example JSON output", since it's fairly standard

daltonbohning commented on src/tests/ftest/util/support_test_base.py:
```diff
@@ -0,0 +1,142 @@
+        self.target_folder = os.path.join("/tmp", "DAOS_Support_logs")
+
+        # make the custom log dir on node (clients or servers)
+        mkdir_cmd = "mkdir -p {}".format(self.custom_log_dir)
+        results = run_pcmd(hosts=self.log_hosts, command=mkdir_cmd)
+        for result in results:
+            if result["exit_status"] != 0:
+                self.fail("Failed to create the custom log dir {} ".format(result))
```
Better to use `run_remote` (newer, better error handling) than `pcmd`
```suggestion
        mkdir_cmd = "mkdir -p {}".format(self.custom_log_dir)
        result = run_remote(self.log, self.log_hosts, mkdir_cmd)
        if not result.passed:
            self.fail("Failed to create the custom log dir {} ".format(result))
```

daltonbohning commented on src/tests/ftest/util/support_test_base.py:
```diff
@@ -0,0 +1,142 @@
+
+        # Create the custom log file on node (clients or servers)
+        create_file = " echo \"{}\" > {}".format(self.custom_log_data, self.custom_log_file)
+        results = run_pcmd(hosts=self.log_hosts, command=create_file)
+        for result in results:
+            if result["exit_status"] != 0:
+                self.fail("Failed to create the custom log file {} ".format(result))
```
```suggestion
        result = run_remote(self,log, self.log_hosts, create_file)
        if not result.passed:
            self.fail("Failed to create the custom log file {} ".format(result))
```

daltonbohning commented on src/tests/ftest/util/support_test_base.py:
```diff
@@ -0,0 +1,142 @@
+        self.custom_log_data = now.strftime("%d/%m/%Y %H:%M:%S")
+
+        # Create the custom log file on node (clients or servers)
+        create_file = " echo \"{}\" > {}".format(self.custom_log_data, self.custom_log_file)
```
Could just use single quotes
```suggestion
        create_file = " echo '{}' > '{}'".format(self.custom_log_data, self.custom_log_file)
```

daltonbohning commented on src/tests/ftest/util/support_test_base.py:
```diff
@@ -0,0 +1,142 @@
+        read_filedata = "find {}  -name {} | xargs cat".format(
+            self.extract_dir, os.path.basename(self.custom_log_file))
+
+        results = run_pcmd(hosts=self.log_hosts, command=read_filedata)
+        for result in results:
+            if result["exit_status"] != 0:
+                self.fail("Failed to read the custom log file {} ".format(result))
+
+            if self.custom_log_data not in result["stdout"]:
+                self.fail("Expected custom_log_data {} not found in log file {}"
+                          .format(self.custom_log_data, result))
+
```
```suggestion
        result = run_remote(self.log, self.log_hosts, read_filedata)
        if not result.passed:
            self.fail("Failed to read the custom log file {} ".format(result))

        for output in result.all_stdout.values():
            if self.custom_log_data not in output:
                self.fail("Expected custom_log_data {} not found in log file {}"
                          .format(self.custom_log_data, result))

```

daltonbohning commented on src/tests/ftest/util/support_test_base.py:
```diff
@@ -0,0 +1,142 @@
+                self.fail("Expected custom_log_data {} not found in log file {}"
+                          .format(self.custom_log_data, result))
+
+    def extract_logs(self, tar_gz_filename):
+        """Extract the logs files which are in collected archive.
```
I'd recommend not using `self.*` references in this function, and instead creating a general purpose function that takes arguments. @phender probably has an opinion here

daltonbohning commented on src/tests/ftest/util/support_test_base.py:
```diff
@@ -0,0 +1,142 @@
+        Args:
+            tar_gz_filename (str): Log archive File name
+
+        Returns:
+            str: Error message if test steps fails, None in case of success
```
I'd recommend just raising an exception instead of returning an error string. Exceptions that aren't caught by a test are implicit test failures

daltonbohning commented on src/tests/ftest/util/support_test_base.py:
```diff
@@ -0,0 +1,142 @@
+        """
+        # Create the new extract directory
+        cmd = "mkdir -p {}".format(self.extract_dir)
+        results = run_pcmd(hosts=self.log_hosts, command=cmd)
```
better to use run_remote

daltonbohning commented on src/tests/ftest/util/support_test_base.py:
```diff
@@ -0,0 +1,142 @@
+
+        # Extract The tar.gz file to newly created directory
+        cmd = "tar -xf {} -C {}".format(tar_gz_filename, self.extract_dir)
+        results = run_pcmd(hosts=self.log_hosts, command=cmd)
```
better to use run_remote

daltonbohning commented on src/tests/ftest/util/support_test_base.py:
```diff
@@ -0,0 +1,142 @@
+        # Verify server log files are collected.
+        for log_file in log_files:
+            list_file = "ls -lsaRt {} | grep {}".format(self.extract_dir, log_file)
+            results = run_pcmd(hosts=self.hostlist_servers, command=list_file)
```
better to use run_remote

daltonbohning commented on src/tests/ftest/util/support_test_base.py:
```diff
@@ -0,0 +1,142 @@
+    def validate_server_log_files(self):
+        """Verify all the server logs files are collected and part of archive.
+
+        Returns:
+            str: Error message if test steps fails, None in case of success
```
recommend instead raising exception

daltonbohning commented on src/tests/ftest/control/daos_agent_support_collect_log.py:
```diff
@@ -0,0 +1,53 @@
+        # Verify the custom log file collected on each clients.
+        self.verify_custom_log_data()
+
+        # Clean up the log file created during test execution
+        self.cleanup_support_log(self.target_folder)
```
This won't run if the test fails. I'll let @phender (more familiar) comment on this

daltonbohning replied:
We could follow this pattern:
https://github.com/daos-stack/daos/blob/652fff38ef0677303669ee9c840f77279027f65a/src/tests/ftest/util/nvme_utils.py#L68
https://github.com/daos-stack/daos/blob/652fff38ef0677303669ee9c840f77279027f65a/src/tests/ftest/util/nvme_utils.py#L79-L97

By changing `self.cleanup_support_log` to return a list of errors instead of calling `self.fail`, and then right after calling `support_collect_log`, register the cleanup with:
`self.register_cleanup(self.reset_fault_device, log_dir=self.target_folder)`

This cleanup will be ran in teardown, which accounts for test failures and timeouts

daltonbohning commented on src/tests/ftest/util/support_test_base.py:
```diff
@@ -0,0 +1,142 @@
+        folders = [log_dir, self.extract_dir]
+        for folder in folders:
+            delete_cmd = "sudo rm -rf {}*".format(folder)
+            results = run_pcmd(hosts=self.log_hosts, command=delete_cmd)
```
better to use run_remote

ravalsam replied:
Done

daltonbohning commented on src/tests/ftest/util/support_test_base.py:
```diff
@@ -0,0 +1,142 @@
+        Args:
+            log_dir (str): Name of the log directory to be removed
+
+        Returns:
+            str: Error message if test steps fails, None in case of success
```
recommend instead raising exception

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/28/execution/node/1190/log

ravalsam commented:
Recent PR single test failed because of an existing issue:

osa/online_drain.py - DAOS-13054

mjmac commented on src/control/common/file_utils.go:
```diff
@@ -257,3 +258,73 @@ func FindBinary(binName string) (string, error) {
+
+// Check if file or directory that starts with . which is hidden
+func IsHidden(filename string) bool {
+	if filename[0:1] == "." {
```
This will panic if `filename` is empty. Should probably add a `filename != "" &&` before this.

mjmac commented on src/control/lib/control/support.go:
```diff
@@ -0,0 +1,69 @@
+		TargetFolder string
+		TargetHost   string
+		ExtraLogsDir string
+		JsonOutput   bool
```
Why does the server need to know about this client-side display option?

ravalsam replied:
We are passing the same option to dmg log collection. Admin has both the option to collect the logs with/without JSON. And support engineer also prefer the JSON in case we have a lot of data to analyze better.

mjmac reviewed (commented):
I checked out the PR to play around with it. When I ran it on my test node, I got the following error:
```
Support CollectLog: Calling Log Function Enum: 11, And Cmd/Log:
DEBUG 14:18:37.523561 interceptors.go:243: gRPC request: *ctl.CollectLogReq (TargetFolder:"/tmp/daos_support_server_logs"  TargetHost:"wolf-61"  LogFunction:11)
ERROR: gRPC handler for *ctl.CollectLogReq failed: Error running command /usr/bin/sh -c 'rsync -av --blocking-io /tmp/daos_support_server_logs/wolf-61 wolf-61:/tmp/daos_support_server_logs ' with exit status 127: exit status 127 (elapsed: 3.601236ms)
```

Exit status 127 is "command not found". I do have rsync on my system, not sure what the issue is.

ravalsam commented:
> I checked out the PR to play around with it. When I ran it on my test node, I got the following error:
> 
> ```
> Support CollectLog: Calling Log Function Enum: 11, And Cmd/Log:
> DEBUG 14:18:37.523561 interceptors.go:243: gRPC request: *ctl.CollectLogReq (TargetFolder:"/tmp/daos_support_server_logs"  TargetHost:"wolf-61"  LogFunction:11)
> ERROR: gRPC handler for *ctl.CollectLogReq failed: Error running command /usr/bin/sh -c 'rsync -av --blocking-io /tmp/daos_support_server_logs/wolf-61 wolf-61:/tmp/daos_support_server_logs ' with exit status 127: exit status 127 (elapsed: 3.601236ms)
> ```
> 
> Exit status 127 is "command not found". I do have rsync on my system, not sure what the issue is.

I will try and verify.

ravalsam commented:
> > I checked out the PR to play around with it. When I ran it on my test node, I got the following error:
> > ```
> > Support CollectLog: Calling Log Function Enum: 11, And Cmd/Log:
> > DEBUG 14:18:37.523561 interceptors.go:243: gRPC request: *ctl.CollectLogReq (TargetFolder:"/tmp/daos_support_server_logs"  TargetHost:"wolf-61"  LogFunction:11)
> > ERROR: gRPC handler for *ctl.CollectLogReq failed: Error running command /usr/bin/sh -c 'rsync -av --blocking-io /tmp/daos_support_server_logs/wolf-61 wolf-61:/tmp/daos_support_server_logs ' with exit status 127: exit status 127 (elapsed: 3.601236ms)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Exit status 127 is "command not found". I do have rsync on my system, not sure what the issue is.
> 
> I will try and verify.

Updated the code, please try now. 
For some reason, exec.command() with separate run() was having the problem. I updated the code with exec.Command("sh", "-c", cmd).Output() and tested locally, and it's working fine now.

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/30/execution/node/174/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daltonbohning commented on src/tests/ftest/control/dmg_support_collect_log.py:
```diff
@@ -0,0 +1,51 @@
+        # Run dmg support collect-log with --extra-logs-dir,
+        # copy logs to folder with command option --target-folder
+        # Enable archive mode.
+        result = self.dmg.support_collect_log(extra_logs_dir=self.custom_log_dir,
+                                              target_folder=self.target_folder,
+                                              archive=True)
+        status = result["status"]
+        self.assertEqual(status, 0, "bad return status from dmg support collect-log")
```
`dmg.support_collect_log` uses `_get_json_result`, which raises an exception by default unless you set `dmg.exit_status_exception = False`. So you'll probably never get a result where status is not 0

ravalsam replied:
Ok, so result.passed is preferred check instead of status?

daltonbohning replied:
`result.passed` is for when using `run_remote`.
`dmg.support_collect_log` uses `_get_json_result`, which returns a dictionary of the json result. But `_get_json_result` will raise an exception if `result["status"] != 0`. Which means the `self.assertEqual(status, 0, ...)` will never happen, because `_get_json_result` would have already raised an exception. You could just remove the status check here :)

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/32/execution/node/138/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daltonbohning reviewed (commented):
ftest LGTM

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daltonbohning reviewed (commented):
ftest LGTM. Doesn't appear to be any changes since my last review.
FYI to be careful with force pushing, since it loses some history for reviewers
> [ravalsam](https://github.com/ravalsam) [force-pushed](https://github.com/daos-stack/daos/compare/365ed5a66ac0bf971ea3f161aea4aead6e1eca85..3e8afeafced28693ad7389faec5a53ef0a312370) the samirrav/Support/DAOS-10625-Final branch from [365ed5a ](https://github.com/daos-stack/daos/commit/365ed5a66ac0bf971ea3f161aea4aead6e1eca85)to [3e8afea ](https://github.com/daos-stack/daos/commit/3e8afeafced28693ad7389faec5a53ef0a312370)
[4 days ago](https://github.com/daos-stack/daos/pull/11094#event-9181006994)

ravalsam commented:
> ftest LGTM. Doesn't appear to be any changes since my last review. FYI to be careful with force pushing, since it loses some history for reviewers
> 
> > [ravalsam](https://github.com/ravalsam) [force-pushed](https://github.com/daos-stack/daos/compare/365ed5a66ac0bf971ea3f161aea4aead6e1eca85..3e8afeafced28693ad7389faec5a53ef0a312370) the samirrav/Support/DAOS-10625-Final branch from [365ed5a ](https://github.com/daos-stack/daos/commit/365ed5a66ac0bf971ea3f161aea4aead6e1eca85)to [3e8afea ](https://github.com/daos-stack/daos/commit/3e8afeafced28693ad7389faec5a53ef0a312370)
> > [4 days ago](https://github.com/daos-stack/daos/pull/11094#event-9181006994)

This was force pushed as I was using the amd/readdir-docs locally as I am still on 8.5 and by mistake that was also merged when I merged the master. So I force pushed and removed that patch.

daltonbohning reviewed (commented):
ftest LGTM but I'm not comfortable with the control code

ravalsam commented:
> ftest LGTM but I'm not comfortable with the control code

Control plane team is listed as reviewer and they will review it again. Your approval requested for Python test code only.

phender commented on src/tests/ftest/util/support_test_base.py:
```diff
@@ -0,0 +1,134 @@
+        server_custom_log = os.environ['AVOCADO_TESTS_COMMON_TMPDIR']
+        self.custom_log_dir = os.path.join(server_custom_log, folder_name)
+        self.custom_log_file = os.path.join(self.custom_log_dir, "Custom_File")
+        self.target_folder = os.path.join("/tmp", "DAOS_Support_logs")
```
We should also use either `self.base_test_dir` or `self.test_dir` here for proper test cleanup and portability.

phender reviewed (changes requested):
The test framework ensures that the `self.base_test_dir` directory (typically set to `/var/tmp/daos_testing`) exists for the duration of all the test variants defined and the `self.test_dir` directory (`self.base_test_dir/self.test_id`) exists for the duration of a single test variant.  These directories are guaranteed to exist on each test host.  I'd suggest using one of these directory for creating or storing files.

If you'd like to collect any files generated by the test after all the test variants have completed, then I'd suggest using the `self.base_test_dir` directory, and naming the files starting with the `self.test_id` - similar to what we do in https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/apricot/apricot/test.py#L1616.  In this case you wouldn't need to define or call `cleanup_support_log(log_dir)` as launch.py will remove and recreate the `self.base_test_dir` directory in-between running test files.

mjmac reviewed (commented):
Changes since last review LGTM. Will re-review once @phender's request is addressed.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/38/execution/node/1033/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/39/execution/node/383/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/40/execution/node/512/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/43/execution/node/1034/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/45/execution/node/1236/log

mjmac reviewed (dismissed):
While this is a huge PR, it seems low risk in that the changes are adding new functionality rather than changing existing code. I'm in support of getting it landed so that we can continue to iterate on it.

kjacque commented on src/control/lib/support/log.go:
```diff
@@ -0,0 +1,733 @@
+}
+
+// Get the system hostname
+func GetHostName() (string, error) {
```
not blocking on this, but there is a Go function `os.Hostname()` that can fetch the hostname without calling out to the shell.

ravalsam replied:
Thanks for review. I will test and updated when will work on second iteration. 

If I recall, I was trying to use that os.Hostname() but it was giving the full name including the domain name. So another option is to use the function and remove the domain from it.

daosbuild1 commented:
Test stage Functional on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11094/47/execution/node/1114/log
